src/constraint.rs

```diff
@@ -1074,3 +1110,165 @@ impl ConstraintValidator for ElementConstraint {
         Ok(())
     }
 }
+
+/// Implements the `annotations` constraint
+/// [annotations]: https://amzn.github.io/ion-schema/docs/spec.html#annotations
+#[derive(Debug, Clone, PartialEq)]
+pub struct AnnotationsConstraint {
+    annotations_modifiers: Vec<AnnotationModifier>,
+    annotations: Vec<Annotation>,
+}
+
+impl AnnotationsConstraint {
+    pub fn new(
+        annotations_modifiers: Vec<AnnotationModifier>,
+        annotations: Vec<Annotation>,
+    ) -> Self {
+        Self {
+            annotations_modifiers,
+            annotations,
+        }
+    }
+
+    pub fn validate_ordered_annotations(
```

Comment:
Thanks for explaining. I see that now. 

minor suggestion—Since you are storing the resolved "requiredness" on each `Annotation`, then it is probably not necessary to model it in the `AnnotationsConstraint`.